Adds ExtensionType for Parquet geospatial WKB arrays

BlakeOrth opened the pull request:
# Which issue does this PR close?

This does not fully close, but is an initial component of:
 - https://github.com/apache/arrow-rs/issues/8717

# Rationale for this change

To keep PR size digestible, this implements the bi-directional extension type metadata parsing for Parquet geospatial arrays. An actual array type that allows users to easily interact with the data can come as a follow-on PR.

# What changes are included in this PR?

 - Implements the ExtensionType for Parquet geospatial logical types
 - Adds GeoArrow compatible Metadata type to hold geospatial metadata
 - Adds basic tests around geospatial metadata serialization/deserialization
 - Integrates geospatial logical type parsing into schema extensions

# Are these changes tested?

Yes. Targeted unit tests for the metadata and associated parsing have been implemented. Higher level tests that show a full round-trip to and from arrays can be included in the array implementation PR.

# Are there any user-facing changes?

Yes. All new public items have been documented, and no API breaking changes have been made.

##
cc @alamb @paleolimbot @kylebarron

kylebarron commented on parquet-geospatial/src/types.rs:
```diff
@@ -0,0 +1,393 @@
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::error::Result;
```
I think in general we prefer using the smallest possible dependency crate surface to minimize compile times. So instead of adding the full `arrow` dependency, it's preferred to use `arrow_schema` instead

BlakeOrth replied:
I'm having trouble finding this in `arrow_schema`, can someone point me to it?

https://docs.rs/arrow-schema/57.1.0/arrow_schema/index.html?search=Result

The only similar thing I've found is a private type alias here:
 https://github.com/apache/arrow-rs/blob/bab30ae3d61509aa8c73db33010844d440226af2/arrow-schema/src/datatype_parse.rs#L29

##
If we want to avoid the dependency altogether there aren't that many `Result<T>` uses here and I can easily type alias or use `std::Result<T, ArrowError>` for those as well.

alamb commented on parquet-geospatial/src/types.rs:
```diff
@@ -0,0 +1,393 @@
+
+/// Well-Known Binary (WKB) [`ExtensionType`] for geospatial data.
+///
+/// Represents the canonical Arrow Extension Type for storing GeoArrow data.
```
Maybe it is worth a link to GeoArrow here https://github.com/geoarrow/geoarrow

BlakeOrth commented:
@alamb @paleolimbot Thank you both for the thorough review! I believe the most recent commit addresses all of the feedback for improvements here. Please let me know if there's any additional feedback for improvements here!

@paleolimbot I believe the geospatial extension test additions are what you were looking for, but let me know if that's not what you had in mind and I'm happy to change those.

alamb commented on parquet-geospatial/src/types.rs:
```diff
@@ -0,0 +1,393 @@
+}
+
+impl ExtensionType for WkbType {
+    const NAME: &'static str = "geoarrow.wkb";
```
> > I've probably already rocked the boat enough
> 
> Loving the rocking!

Ditto

BlakeOrth commented:
@alamb It looks like my most recent test additions to `parquet/tests/geospatial.rs` broke the "standard" build because it's now using imports from the feature gated `parquet-geospatial` crate (that'll will teach me to use the `--all-features` flag). Would it be ok to gate this entire test behind that feature or maybe I should move the affected tests into the `parquet-geospatial` crate?

kylebarron commented on parquet-geospatial/src/types.rs:
```diff
@@ -0,0 +1,394 @@
+    /// Note: Common lon/lat CRS representations (EPSG:4326, OGC:CRS84) are canonicalized
+    /// to `None` during serialization to match Parquet conventions.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub crs: Option<serde_json::Value>,
```
Likewise I think having a plain `serde_json::Value` here leads to opportunities for invalid formatting of CRS. I don't know whether we expect this code to validate input or whether it's expected for validation to be somewhere else down the pipeline. 

In particular, in geoarrow-schema I split up the underlying `crs` and `crs_type`: https://github.com/geoarrow/geoarrow-rs/blob/f1deee4549760c7cd8476511525115c185515272/rust/geoarrow-schema/src/crs.rs#L35-L62

Or is the encoding in the Parquet geometry spec not the same as the encoding in GeoArrow?

BlakeOrth commented on parquet-geospatial/src/types.rs:
```diff
@@ -0,0 +1,394 @@
+    pub crs: Option<serde_json::Value>,
+    /// The edge interpolation algorithm of the [`WkbType`], if present.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub algorithm: Option<String>,
```
Yes, due to the dependency graph we would have to effectively duplicate this enum within the `parquet-geospatial` crate:
https://github.com/apache/arrow-rs/pull/8943/files#diff-77b1a58271f532bba48f784fc34eefd52639735547f9e32187dbfbf485ae179dL998

I agree that we already validate the value on each leg of the round-trip, so I don't think we're likely to get unexpected values from the raw string interpretation. It seems like it's mostly a matter of how we'd like to interact with this downstream. An enum results in some duplicated code, but it's probably a bit more idiomatic from a user's perspective. @kylebarron I suspect `geoarrow-rs` is likely to be the primary consumer of this metadata, so I'd probably generally defer to you on what you think the best way to interact with this metadata is.

BlakeOrth replied:
I swapped this over to an enum and implemented the bi-directional conversion (gated behind the geospatial feature)

BlakeOrth commented on parquet-geospatial/src/types.rs:
```diff
@@ -0,0 +1,394 @@
+            algorithm: self.0.algorithm.clone(),
+        };
+
+        serde_json::to_string(&md).ok()
```
I was able to make this all much cleaner and more readable thanks to this suggestion. Thank you!
